### PR TITLE
Benchmarks, part 2

### PR DESCRIPTION
**Pull request by anton-malakhov**
@jeremiedbb Please take a look

### CODE REVIEW

**Comment by jeremiedbb** [comment]
Thanks ! I look take a look tomorrow. I noticed a large amount of auto formatting changes. Can you revert all these changes in the meanwhile ? it will make the review way easier :)

**Comment by anton-malakhov** [comment]
@0just0 Could you please revert the auto-formatting changes?

File: benchmarks/crossvalidation.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
+from sklearn.ensemble import RandomForestClassifier
+from sklearn.model_selection import GridSearchCV, cross_val_score
+
+from .common import Benchmark
+from .datasets import _synth_classification_dataset
+
+
+class CrossValidation_bench(Benchmark):
+    def setup(self, params):
```
Comment (jeremiedbb) [inline benchmarks/crossvalidation.py]: I'd rather split cross_val_score and GridSearchCV. Everywhere else I named the file same as the corresponding scikit-learn module. Here it would be 'model_selection.py'. Then I made a separate class for each class or function we want to benchmark.

It also has the advantage to easily run the benchmark for one class: `asv run -b TheClassName`

File: benchmarks/crossvalidation.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        self.n_jobs = params[0]
+        self.X, self.y = _synth_classification_dataset(n_samples=70000,
+                                                       n_features=200,
+                                                       random_state=0)
+
+    def time_crossval(self, params):
+        clf = RandomForestClassifier(n_estimators=100, random_state=0)
```
Comment (jeremiedbb) [inline benchmarks/crossvalidation.py]: Please also add a peakmem function. This is important for the cross validation tools.

File: benchmarks/datasets.py [hunk 3]
```diff
@@ -83,17 +81,31 @@ def _synth_classification_dataset(n_samples=1000, n_features=10000,
...
         X[X < 2] = 0
         X = sp.csr_matrix(X)
-
     X = MaxAbsScaler().fit_transform(X)
     return X, y
 
 
+@M.cache
+def _decomposition_dataset():
+    """
```
Comment (jeremiedbb) [inline benchmarks/datasets.py:94]: I'd name it `olivetti_faces_dataset` instead. It might also be used for non decomposition benchmarks later.

File: benchmarks/linear_model.py [hunk 4]
```diff
@@ -80,10 +83,62 @@ def peakmem_fit(self, *args):
         ridge.fit(self.X, self.y)
 
 
+class Linear_bench(Benchmark):
+    """
```
Comment (jeremiedbb) [inline benchmarks/linear_model.py:87]: Please name it `LinearRegression_bench`.

**Comment by jeremiedbb** [comment]
Thanks for the new benchmarks classes. I made a few comments.

making MULTICORE a list is nice.

Have you tried running the whole benchmark suite ? I'm afraid setting `repeat = (3,100,120)` makes it take a few hours. I'll think about reducing the sizes of some datasets to achieve reasonable times.

**Comment by ogrisel** [comment]
With respect to stability of the benchmark here is a very interesting resource that we might want to consider:

https://perf.readthedocs.io/en/latest/

in particular, the `perf system tune` command:

https://perf.readthedocs.io/en/latest/cli.html#system-cmd
https://perf.readthedocs.io/en/latest/system.html

**Comment by jeremiedbb** [comment]
Good to merge. Thanks !

**Comment by jeremiedbb** [comment]
Are all the estimators/functions you improved being part of the benchmark suite now ?

**Comment by oleksandr-pavlyk** [comment]
Yes, all estimators/functions that were affected by daal4py patches are covered.  These currently include SVC, PCA (svd_solver = 'full'), LinearRegression, RidgeRegression, KMeans, LogisticRegression/logistic_regression_path with solver='lbfgs', as well as cosine and correlation distances.
